### ISSUE TITLE
New-MgGroup - Randomly occurring Exception during group creation - Microsoft.Graph v. 2.0.0-rc3, Powershell 5.1

### ISSUE THREAD

**Issue by Mpmdg** [microsoftgraph/msgraph-sdk-powershell#2064]
Reproduction steps:

``` powershell
$requiredModules = @(
  'Microsoft.Graph.Authentication'
  'Microsoft.Graph.Groups'
)

$jobs = @()

1..50 | ForEach-Object {
$jobs += Start-RSJob -Name "Add group: {$_}" -ModulesToImport $requiredModules -Throttle 10 -ScriptBlock {
    $grp = New-MgGroup -DisplayName "new-mggroup-thread-test" -MailEnabled:$false -MailNickname 'unused' -SecurityEnabled
    if ($null -eq $grp)
    {
      return "Null received from New-MgGroup!" 
    } 
    else
    {
      return "Group {$_} added successfully."
    }
  }
}

while ($jobs.Completed -contains $False) {
    Start-Sleep -Seconds 1
}

$jobs | Wait-RSJob | Receive-RSJob
```
Result:

Sometimes when creating groups in parallel, a few requests fail with errors:
``` powershell
> $jobs | Wait-RSJob | Receive-RSJob

One or more errors occurred.
Null received from New-MgGroup!
Group 2 added successfully.
Group 3 added successfully.
Group 4 added successfully.

> $jobs
Id       Name                 State           HasMoreData  HasErrors    Command
--       ----                 -----           -----------  ---------    -------
117      Add group: {1}       Completed       True         True         ...
118      Add group: {2}       Completed       True         False        ...
119      Add group: {3}       Completed       True         False        ...
120      Add group: {4}       Completed       True         False        ...
```

The exception message retrieved from the job object:
``` powershell
Message        : Collection was modified; enumeration operation may not execute.
Data           : {}
InnerException :
TargetSite     : Void ThrowInvalidOperationException(System.ExceptionResource)
StackTrace     :    at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
                    at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
                    at System.Net.Http.Headers.HttpHeaders.<GetEnumerator>d__20.MoveNext()
                    at Microsoft.Graph.PowerShell.HttpMessageLogFormatter.<CloneAsync>d__0.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Graph.PowerShell.HttpMessageLogFormatter.<GetHttpRequestLogAsync>d__1.MoveNext()
HelpLink       :
Source         : mscorlib
HResult        : -2146233079
```

----------
The errors occur randomly - I once ran 100 parallel requests of which all succeed, and the other time I had 2 errors in 50 requests.

Should the Powershell Graph be considered as thread safe, or is multithreading not officially supported?

---------

Update:

I was able to reproduce the same error using the built-in `Start-Job | Wait-Job | Receive-Job` commands.

I am however __unable__ to reproduce this issue on Powershell 7.3, using neither `Start-Job`, nor `ForEach-Object -Parallel`.

--------

Update 2:

While I initially reported that the issue is caused by parallel execution, I recently discovered that it randomly occurs during synchronous command calls as well.

The following code has a high probability of producing the exception mentioned above:
``` powershell
1..100 | ForEach-Object {
  New-MgGroup -DisplayName "new-mggroup-test" -MailEnabled:$false -MailNickname 'unused'
}
```

**Comment by jberezanski-mdg** [comment]
To clarify - we are using the [PoshRSJob](https://github.com/proxb/PoshRSJob) module (version 1.7.4.4) to run the operations in parallel. Until now, we have been creating the groups using the AzureAD module with no issues.

**Comment by peombwa** [comment]
The stack trace points to this possibly being a duplicate of https://github.com/microsoftgraph/msgraph-sdk-powershell/issues/1916, which got fixed in v2-rc3. The bug was traced to how the http headers were being handled by the underlaying `WinHttpHandler` in .NET fx (PowerShell 5.1).

Could you share the following details to help us better repro the issue:
1. Which version of PowerShell are you using? Please provide the output of `$PSVersionTable`.
2. Does the error occur in both PowerShell 5.1 and PowerShell 7.* or is it just isolated to PowerShell 5.1?
3. Which version of the SDK are you using? Please provide the out of `Get-Module Microsoft.Graph.*`.

**Comment by jberezanski-mdg** [comment]
We are using:
- PowerShell 5.1
- Microsoft.Graph.* version 2.0.0-rc3

(I'm away from computer at the moment so I'm unable to provide the command output. I'll be able to post the evidence after the weekend.)

The stack trace points to the `HttpMessageLogFormatter.CloneAsync()` method added specifically in #2028 as a fix to #1916. So it seems the fix was not 100% sufficient.
```
                    at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
                    at System.Net.Http.Headers.HttpHeaders.<GetEnumerator>d__20.MoveNext()
                    at Microsoft.Graph.PowerShell.HttpMessageLogFormatter.<CloneAsync>d__0.MoveNext()
```

It looks like something already modifies `originalRequest.Headers` while `CloneAsync()` is executing.

**Comment by Mpmdg** [comment]
I'll confirm what @jberezanski-mdg has stated above:

1. The output of $PSVersionTable:
``` powershell
Name                           Value
----                           -----
PSVersion                      5.1.19041.2673
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.19041.2673
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

2. I tried to reproduce the error using PS 7.*, but the error did not occur even once, even though I've created about ~2k groups in a ForEach loop. I am able to reliably reproduce it using PS 5.1 (20 iterations of ForEach is often more than enough).

3. Below are the outputs of Get-Module, as well as Get-InstalledModule, to confirm that 2.0.0-rc3 is the only version installed:
``` powershell
> Get-Module Microsoft.Graph*

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.0.0      Microsoft.Graph.Applications        {Add-MgApplicationKey, Add-MgApplicationPassword, Add-MgSe...
Script     2.0.0      Microsoft.Graph.Authentication      {Add-MgEnvironment, Connect-MgGraph, Disconnect-MgGraph, G...
Script     2.0.0      Microsoft.Graph.Beta.Groups         {Add-MgBetaGroupDriveListContentTypeCopy, Add-MgBetaGroupD...
Script     2.0.0      Microsoft.Graph.Beta.Identity.Di... {Complete-MgBetaDirectoryImpactedResource, Complete-MgBeta...
Script     2.0.0      Microsoft.Graph.Groups              {Add-MgGroupDriveListContentTypeCopy, Add-MgGroupDriveList...
Script     2.0.0      Microsoft.Graph.Identity.Directo... {Confirm-MgContactMemberGroup, Confirm-MgContactMemberObje...
Script     2.0.0      Microsoft.Graph.Users               {Get-MgUser, Get-MgUserCount, Get-MgUserCreatedObject, Get...

> Get-InstalledModule Microsoft.Graph* | Group-Object -Property Version

Count Name                      Group
----- ----                      -----
   77 2.0.0-rc3                 {@{Name=Microsoft.Graph.Applications; Version=2.0.0-rc3; Type=Module; Description=Mi...
```

**Comment by peombwa** [comment]
Thanks for detailed repro steps!

I'm able to repro the issue on my end. Our preliminary investigation points to a thread safety issue in .NET Framework (4.x, PowerShell 5.1 runtime) when reading content headers concurrently. The runtime attempts to remove invalid/duplicate header values when enumerating the content headers in the clone method which results in the "collection was modified" error. See https://github.com/dotnet/runtime/issues/61798 for more details.

[Locks have since been added in .NET 7](https://github.com/dotnet/runtime/pull/68115) to support concurrent reads. This explains why the bug is not only present in PowerShell 5.1. As a workaround, you can use PowerShell 7 to run your scripts as we work on remedy for .NET 4.x (PowerShell 5.1).

**Comment by jberezanski-mdg** [comment]
Thank you for the explanation. This issue actually motivated us to finally move forward with the long-postponed migration of our devops automation to PS 7 (the only thing holding us back was the AzureAD module, which we have now replaced with Mg) and so far it looks promising, so we may stop caring about PS 5.1 support soon.

**Comment by peombwa** [comment]
The bug fix shipped as part of [v2.0.0 GA release](https://github.com/microsoftgraph/msgraph-sdk-powershell/releases/2.0.0). Please upgrade to the latest version and let us know if the issue is still present.

**Comment by Mpmdg** [comment]
I confirm than I am unable to reproduce the error with the 2.0.0 release version.

Thank you for fixing the issue and keeping us updated on this matter!
